pallets/spin-polkadot/src/lib.rs

```diff
@@ -0,0 +1,257 @@
+// mod tests;
+
+// #[cfg(feature = "runtime-benchmarks")]
+// mod benchmarking;
+
+pub mod weights;
+
+#[frame::pallet]
+pub mod pallet {
+    use frame::{prelude::*, runtime::types_common::BlockNumber, traits::Header};
+    use polkadot_parachain_primitives::primitives::HeadData;
+    use sp_consensus_grandpa::Commit;
+    use sp_runtime::Vec;
+
+    #[derive(PartialEq, Eq, Clone, Encode, Decode, TypeInfo, RuntimeDebug)]
+    pub struct PersistedValidationData<H = H256, N = BlockNumber> {
+        /// The parent head-data.
+        pub parent_head: HeadData,
+        /// The relay-chain block number this is in the context of.
+        pub relay_parent_number: N,
+        /// The relay-chain block storage root this is in the context of.
+        pub relay_parent_storage_root: H,
+        /// The maximum legal size of a POV block, in bytes.
+        pub max_pov_size: u32,
+    }
+
+    /// The inherent data that is passed by the collator to the parachain runtime.
+    #[derive(Encode, Decode, RuntimeDebug, Clone, PartialEq, TypeInfo)]
+    pub struct FastchainInherentData {
+        pub validation_data: PersistedValidationData,
+        /// A storage proof of a predefined set of keys from the relay-chain.
+        ///
+        /// Specifically this witness contains the data for:
+        ///
+        /// - the current slot number at the given relay parent
+        /// - active host configuration as per the relay parent,
+        /// - the relay dispatch queue sizes
+        /// - the list of egress HRMP channels (in the list of recipients form)
+        /// - the metadata for the egress HRMP channels
+        pub relay_chain_state: sp_trie::StorageProof,
+        // /// Downward messages in the order they were sent.
+        // pub downward_messages: Vec<InboundDownwardMessage>,
+        // /// HRMP messages grouped by channels. The messages in the inner vec must be in order they
+        // /// were sent. In combination with the rule of no more than one message in a channel per block,
+        // /// this means `sent_at` is **strictly** greater than the previous one (if any).
+        // pub horizontal_messages: BTreeMap<ParaId, Vec<InboundHrmpMessage>>,
+    }
+
+    #[derive(Clone, Encode, Decode, RuntimeDebug, PartialEq, Eq, TypeInfo)]
+    pub struct GrandpaJustification<H: Header> {
+        pub round: u64,
+        pub commit: Commit<H>,
+        pub votes_ancestries: Vec<H>,
+    }
+
+    #[derive(Encode, Decode, RuntimeDebug, Clone, PartialEq, TypeInfo)]
+    pub struct AliveMessageProof<H: Header> {
+        pub fastchain_inherent_data: FastchainInherentData,
+        pub grandpa_justification: GrandpaJustification<H>,
+    }
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+        type WeightInfo: crate::weights::WeightInfo;
+
+        #[pallet::constant]
+        type TimeoutBlocks: Get<BlockNumber>;
```

Comment:
Are those parachain blocks? What Thunderella uses to measure time?

Reply:
Yes, it is parachain blocks. Thunderella measures time in blocks [as well](https://eprint.iacr.org/2017/913.pdf).
<img width="959" alt="Screenshot 2025-03-06 at 16 48 01" src="https://github.com/user-attachments/assets/a203779f-4609-445c-9f6b-c0c171328bc9" />

Reply:
Seems like Thunderella uses underlying blockchain block time which should be Polkadot in our case.